File: src/Library/SubscriptionResponseProcessor.vb
Imports System.Threading

''' <summary>
''' Processes responses for subscription RCP methods.
''' </summary>
''' <remarks></remarks>
Friend Class SubscriptionResponseProcessor
    Inherits Pixa.Soundbridge.Library.ResponseProcessorBase

    Private _eventRaiser As Action(Of String)
    Private _receivedSubAck As Boolean

    ''' <summary>
    ''' Instantiates a new instance of ResponseProcessorBase for the specified SoundbridgeClient and EventWaitHandle.
    ''' </summary>
    ''' <param name="client"></param>
    ''' <param name="waitHandle"></param>
    ''' <remarks></remarks>
    Public Sub New(ByVal client As SoundbridgeClient, ByVal command As String, ByVal waitHandle As EventWaitHandle, ByVal eventRaiser As Action(Of String))
        MyBase.New(client, command, waitHandle)
        _eventRaiser = eventRaiser
    End Sub

    ''' <summary>
    ''' Processes the specified response line.
    ''' </summary>
    ''' <param name="response"></param>
    ''' <remarks></remarks>
    Public Overrides Sub Process(ByVal response As String)
        If _receivedSubAck Then
            _eventRaiser(response)
        Else
            AddResponse(response)
            _receivedSubAck = True
            WaitHandle.Set()
        End If
    End Sub

    Public Overrides Sub PostProcess()
        If Response.Length = 0 Then ExceptionHelper.ThrowCommandTimeout(Command)
    End Sub
End Class

File: src/Pixa.Soundbridge/Client/SubscriptionResponseProcessor.cs
using System;
using System.Threading;

namespace Pixa.Soundbridge.Client
{

    /// <summary>
    /// Processes responses for subscription RCP methods.
    /// </summary>
    /// <remarks></remarks>
    internal class SubscriptionResponseProcessor : ResponseProcessorBase
    {
        private Action<string> _eventRaiser;
        private bool _receivedSubAck;

        /// <summary>
        /// Instantiates a new instance of ResponseProcessorBase for the specified SoundbridgeClient and EventWaitHandle.
        /// </summary>
        /// <param name="client"></param>
        /// <param name="waitHandle"></param>
        /// <remarks></remarks>
        public SubscriptionResponseProcessor(TcpSoundbridgeClient client, string command, EventWaitHandle waitHandle, Action<string> eventRaiser) : base(client, command, waitHandle)
        {
            _eventRaiser = eventRaiser;
        }

        /// <summary>
        /// Processes the specified response line.
        /// </summary>
        /// <param name="response"></param>
        /// <remarks></remarks>
        public override void Process(string response)
        {
            if (_receivedSubAck)
            {
                _eventRaiser(response);
            }
            else
            {
                AddResponse(response);
                _receivedSubAck = true;
                WaitHandle.Set();
            }
        }

        public override void PostProcess()
        {
            if (Response.Length == 0)
                ExceptionHelper.ThrowCommandTimeout(Command);
        }
    }
}